### ISSUE TITLE
Remove hard-coded `Vector` fields and add vectors constructors

### ISSUE THREAD

**Issue by theogf** [JuliaGaussianProcesses/KernelFunctions.jl#299]
Right now kernels like `LinearKernel` have two problems: they can only take `Real` arguments (no way to pass `kernel.c`) as an argument and also they do not allow for different `AbstractVector` types to be stored.
The interested use case is when using GPUs, one cannot do `kernelmatrix` on `CuArrays` without this.

```julia
struct LinearKernel{Tc<:Real} <: SimpleKernel
    c::Vector{Tc}

    function LinearKernel(c::Real)
        @check_args(LinearKernel, c, c >= zero(c), "c ≥ 0")
        return new{typeof(c)}([c])
    end
end
```
should be
```julia
struct LinearKernel{Tc<:Real,Vc<:AbstractVector{<:Tc}} <: SimpleKernel
    c::V
    function LinearKernel(c::V) where {T<:Real, V<:AbstractVector{T}}
        @check_args(LinearKernel, first(c), first(c) >= zero(c), "c ≥ 0")
        return new{T,V}(c)
    end
    function LinearKernel(c::Real)
        @check_args(LinearKernel, c, c >= zero(c), "c ≥ 0")
        C = [c]
        return new{eltype(C),typeof(C)}(C)
    end
end
```

**Comment by devmotion** [comment]
Or we could just not make it a vector at all but a scalar :stuck_out_tongue:

**Comment by devmotion** [comment]
If we stick with vectors (e.g. because of Flux/Zygote), in the example I guess `LinearKernel(c::Real)` could be a simple outer constructor `LinearKernel(c::Real) = LinearKernel([c])`.

**Comment by theogf** [comment]
But then how do I do my inplace updates! :laughing:. Also nobody like scalars! Especially GPUs!

**Comment by devmotion** [comment]
One could eg use ParameterHandling and perform in-place updates of the resulting vector 😉 But then you would still have to go back and reconstruct the kernel.

**Comment by willtebbutt** [comment]
Did we decide that `Ref`s were a bad idea? i.e. we shouldn't be using `Ref(5.0)` doesn't play nicely with Functors.jl?

**Comment by theogf** [comment]
`Ref` is not captured by `Flux.params` AFAICT

**Comment by devmotion** [comment]
I still think one could consider dropping support for Flux.params and switching to scalar parameters instead. Both Ref and Vector seem a bit unnatural.

**Comment by willtebbutt** [comment]
> Ref is not captured by Flux.params AFAICT

:(

> I still think one could consider dropping support for Flux.params and switching to scalar parameters instead. Both Ref and Vector seem a bit unnatural.

I can definitely see the appeal of this, but I would really like this ecosystem to be compatible with the Flux one e.g. so that you can map the inputs to a GP through a DNN defined using Flux. I'm not saying that this is necessarily a good idea in general, see [1], but I'd at least like to avoid ruling it out.

If we didn't use the Functors API, is there another way to interop nicely with Flux?

[1] - https://arxiv.org/abs/2102.12108

**Comment by theogf** [comment]
> One could eg use ParameterHandling and perform in-place updates of the resulting vector  But then you would still have to go back and reconstruct the kernel.

You mentionned ParameterHandling and it looks very nice and it seems the only requirement on our side would be to implement the `flatten` approach. However that means any additional operations on the kernels would need to follow the ParameterHandling approach no?

**Comment by willtebbutt** [comment]
Not necessarily. You _can_ implemented `flatten` for all of the things in your model, but my preferred approach is to use ParameterHandling.jl to make it easy to write a function which builds your model.

For example, [see here](https://github.com/JuliaGaussianProcesses/TemporalGPs.jl#learning-kernel-parameters-with-optimjl-parameterhandlingjl-and-zygotejl).

This is a different style from the one that you prefer though @theogf .

**Comment by devmotion** [comment]
I haven't thought this through but maybe it is possible to write a custom Flux layer that contains only the parameter vector and a function that builds the kernel:
´``julia
struct KernelLayer{P,F}
    params::P
    model::F
end

Functors.@functor KernelLayer (params,)

(f::KernelLayer)(x, y) = f.model(f.params)(x, y)
...
```

**Comment by theogf** [comment]
> Not necessarily. You _can_ implemented `flatten` for all of the things in your model, but my preferred approach is to use ParameterHandling.jl to make it easy to write a function which builds your model.

I think I got this part. However I wanted to say that I am also happy with the `destructure` approach of `Flux/Functors` as well, which seems "similar" to ParameterHandling.

My point is that if we can have `params, reconstruct = flatten(my_very_complicated_kernel)` I am also happy with that :)

**Comment by willtebbutt** [comment]
> I haven't thought this through but maybe it is possible to write a custom Flux layer that contains only the parameter vector and a function that builds the kernel:

Interesting -- how are you imagining this interacting with Flux @devmotion ? I don't think I'm seeing where you're going with this yet.

**Comment by devmotion** [comment]
As I said, I didn't spend much thought on it. But the main idea would be that instead of forcing all kernels to be mutable we just have one kernel that just wraps the parameters of a possibly complicated kernel in a Flux-compatible way (e.g. as a `Vector`) and includes a function that constructs the kernel from the parameters. Due to the `functor` definition, Flux would only work with and update the parameters.

**Comment by theogf** [comment]
> As I said, I didn't spend much thought on it. But the main idea would be that instead of forcing all kernels to be mutable we just have one kernel that just wraps the parameters of a possibly complicated kernel in a Flux-compatible way (e.g. as a `Vector`) and includes a function that constructs the kernel from the parameters. Due to the `functor` definition, Flux would only work with and update the parameters.

Ok so that's like a complementary idea right? We would need a conversion from **any** kernel to a `KernelLayer` to be then used with any optimization tool?
The conversion could be handled in back-end by ParameterHandling or something handcrafted.

**Comment by devmotion** [comment]
> We would need a conversion from any kernel to a KernelLayer to be then used with any optimization tool?

The main idea is that it only requires the output of `ParameterHandling.flatten` or `Functors.functor` (if one prefers @theogf's approach) or the custom function to build the kernel (apparently @willtebbutt's preferred approach). So there is not any additional support needed.

**Comment by devmotion** [comment]
E.g., in https://juliagaussianprocesses.github.io/KernelFunctions.jl/stable/examples/deep-kernel-learning/ one approach would be to remove the `@functor` definitions (I don't think they are needed currently either, they are defined in KernelFunctions already and not useful for kernels without parameters such as `SqExponentialKernel`) and replace 
```julia
neuralnet = Chain(Dense(1, 3), Dense(3, 2))
k = SqExponentialKernel() ∘ FunctionTransform(neuralnet)
```
with
```julia
neuralnet = Chain(Dense(1, 3), Dense(3, 2))
k = KernelModel(flatten(SqExponentialKernel())...) ∘ FunctionTransform(neuralnet)
```
Then the rest of the example would work in the same way as now and it would be as efficient as now. Clearly, this is a stupid example since `SqExponentialKernel` does not have any parameters and it would be "cooler" if `SqExponentialKernel()` is replaced by a combination of kernels with different parameters. I would suggest to use `@functor` only for such high-level kernels such as `TransformedKernel` (which `k` is an example of) and transforms such as `FunctionTransform` where we want to propagate always to nested elements. Proper kernels, with e.g. real-valued or positive parameters, would have to be casted to a `KernelModel` by users (as mentioned above, a more convenient syntax than ``KernelModel(flatten(...)...)` would be nice) or already be designed as `KernelModel`, and are ignored otherwise by `functor` and hence e.g. `Flux.Params`.

**Comment by theogf** [comment]
Ahhh I think I understand now!
You're saying we provide a `KernelModel` interface containing the parameters and the reconstruction function right?. It is compatible with `Functors`, because we just tell it that the flattened array representation are the trainable parameters.
In this case, I would even argue that we don't have to implement `@functor` only for specific objects (like `FunctionTransform`) as we can just use `flatten`. 

We could default `flatten` to try to use the `functor` API for unknown `struct`s/objects. Such that `flatten(neuralnet)` would just call the `Functors` API
